airflow/providers/salesforce/example_dags/example_tableau_refresh_workbook.py

```diff
@@ -0,0 +1,59 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+This is an example dag that performs a refresh operation on a Tableau Workbook aka Extract. Since this is an
+asynchronous operation we don't know when the operation actually finishes. That's why we have a second task
+that checks exactly that. So that you can perform further operations after the extract has been refreshed.
+"""
+from datetime import timedelta
+
+from airflow import DAG
+from airflow.providers.salesforce.operators.tableau_refresh_workbook import TableauRefreshWorkbookOperator
+from airflow.providers.salesforce.sensors.tableau_job_status import TableauJobStatusSensor
+from airflow.utils.dates import days_ago
+
+DEFAULT_ARGS = {
+    'owner': 'airflow',
+    'depends_on_past': False,
+    'start_date': days_ago(2),
+    'email': ['airflow@example.com'],
+    'email_on_failure': False,
+    'email_on_retry': False
+}
+
+with DAG(
+    dag_id='example_tableau_refresh_workbook',
+    default_args=DEFAULT_ARGS,
+    dagrun_timeout=timedelta(hours=2),
+    schedule_interval=None,
+    tags=['example'],
+) as dag:
+    task_refresh_workbook = TableauRefreshWorkbookOperator(
+        site_id='my_site',
+        workbook_name='MyWorkbook',
+        task_id='refresh_tableau_workbook',
+        dag=dag
+    )
+    task_check_job_status = TableauJobStatusSensor(
```

Comment:
Personally I'm not a big fan of this pattern. I know that this was introduced with the AWS stuff a while ago, but I'd like to have the original Operator to be blocking (maybe make this configurable), instead of always using an Operator/Sensor pair.

For example, if the `TableauRefreshWorkbookOperator` operation fails for some reason. We'd like to retry this. If the exit code is given at the Sensor, the sensor will be retried.

Reply:
> instead of always using an Operator/Sensor pair.

That's where I am using subdags :) ❤️

Reply:
>  I'd like to have the original Operator to be blocking (maybe make this configurable)

So you want it to `sleep` ?

Reply:
Normally the Operator itself would block until it has refreshed. So the behavior of the Sensor is part of the Operator.
For me, having a separate sensor is a bit awkward. For example with the BashOperator, we wait until the process has finished, we don't have a sensor that will block until the process has ended.

Reply:
I could make this configurable so that after the refresh operation has been kicked off It will poke until it succeeds and then the task status will be success.
But this would mean that you can have this bad case where the refresh operation has been kicked off successfully but somehow the poke fails due to a unusual error. When you then clear this task (because it failed) you will kick off another extract refresh.

Reply:
I think having a separate sensor or making it configurable is crucial.

When you perform Extract the processing of the job is "moving" from Airflow worker to Tableau server. There is no reason for the Airflow operator to continue occupying the worker needlessly.

BashOperator is different. The script runs on the Airflow worker resources. It can also be used to connect with another service resulted in waiting time but this is an option unlike the TableauRefreshWorkbookOperator where it always executed on the Tableau server resources.

Think also about a case where many Dags uses TableauRefreshWorkbookOperator at the same time. If all occupy the workers till extract is finished it might cause the entire Airflow cluster to be paralyzed.

I think @feluelle concern about 2nd refresh is a valid one.

Reply:
That depends on how you implement it. If you fetch the and check if it is still refreshing, then you should not re-refresh the workbook. You need to handle different exceptions in a different way.

The implementation of the Operator is a different discussion. Right now this is getting cluttered in the codebase. For GCP we're having a single operator that blocks, while many AWS operators have this Operator/Sensor pattern.

Reply:
My preference would be, making the Operator blocking by default, and also supply the sensor. People can easily use the operator itself if they have only one workbook for example. If they go to more advanced patterns, they can use the sensor pattern. WDYT?

Reply:
I have no strong feelings about what should be the default. From my perspective as long as overwrite is possible it's good enough.

Reply:
@Fokko it really sounds reasonable to me, but I tried to implement it and I don't think it works out nicely... :/

Here is what I have right now:
```python
with TableauHook(self.site_id, self.tableau_conn_id) as tableau_hook:
    workbook = self._get_workbook_by_name(tableau_hook)

    job_id = self._refresh_workbook(tableau_hook, workbook.id)
    if self.blocking:
        from airflow.providers.salesforce.sensors.tableau_job_status import TableauJobStatusSensor
        TableauJobStatusSensor(job_id, self.site_id, self.tableau_conn_id).poke(context)
    return job_id
```

Reply:
And do I wait until it has succeeded or until it has finished?

Reply:
I would `sleep`, similar to here: https://github.com/apache/airflow/blob/aff3a361b4092212c0757f9ce88fa2e40d25d1f4/airflow/providers/apache/druid/hooks/druid.py#L99-L124

Reply:
I am using the Sensor right know with the default mode `poke` means it will `sleep` :)

Reply:
I need to fully test this behaviour first.

Reply:
Seems to work :)